File: scala/anagram/src/main/scala/anagram.scala
class Anagram(base: String) {

  val baseLower = base.toLowerCase
  val baseSorted = baseLower.sorted

  def matches(words: Seq[String]): Seq[String] = {
    words.filter(isAnagram(_))
  }

  def isAnagram(word: String) = {
    val wordLower = word.toLowerCase

    (baseLower != wordLower) && (baseSorted == wordLower.sorted)
  }
}

File: java/anagram/src/main/java/Anagram.java
import java.util.List;
import java.util.stream.Collectors;

public class Anagram {

  private String word;
  private String sorted;

  public Anagram(String word) {
    this.word = word;
    this.sorted = sort(word);
  }

  public List<String> match(List<String> words) {
    return words.stream()
      .filter(this::isAnagram)
      .collect(Collectors.toList());
  }

  private String sort(String word) {
    return word.toLowerCase().chars()
      .sorted()
      .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
      .toString();
  }

  private boolean isAnagram(String candidate) {
    return !this.word.equalsIgnoreCase(candidate) && this.sorted.equals(sort(candidate));
  }

}
